Mathlib/LinearAlgebra/RootSystem/Basic.lean

```diff
@@ -264,34 +137,53 @@ lemma coroot_eq_coreflection_of_root_eq_of_span_eq_top' [CharZero R] [NoZeroSMul
-  apply p.bijectiveRight.injective
-  apply Dual.eq_of_preReflection_mapsTo hk₀ (finite_range root) hsp (hp k) (hs k)
-  · simp [α, β, α', β', sα, sβ, sα', hk, preReflection_apply, hp i, hp j, mul_two,
-      mul_comm (p.toLin α β')]
-    ring -- v4.7.0-rc1 issues
-  · rw [hk, hij]
-    exact (hs i).comp <| (hs j).comp (hs i)
-
-/-- In characteristic zero if there is no torsion and the roots span, if the `i`th reflection of the
-`j`th root is the `k`th root, then the corresponding relationship holds for coroots. See also
-`RootSystem.coroot_eq_coreflection_of_root_eq`. -/
-lemma coroot_eq_coreflection_of_root_eq_of_span_eq_top [CharZero R] [NoZeroSMulDivisors R M]
-    (hsp : span R (range P.root) = ⊤)
-    {i j k : ι} (hk : P.root k = P.reflection i (P.root j)) :
-    P.coroot k = P.coreflection i (P.coroot j) :=
-  coroot_eq_coreflection_of_root_eq_of_span_eq_top' P.toPerfectPairing P.root P.coroot
-    P.coroot_root_two P.mapsTo_preReflection_root hsp hk
+  obtain ⟨l, hl⟩ := hc i (mem_range_self j)
+  rw [← hl]
+  have hkl : (p.toLin.flip (coroot l)) (root k) = 2 := by
+    simp [hl, hk, preReflection_apply, mul_sub, mul_two, β', α, α', β, sα, hp i, hp j]
+    rw [mul_comm (p.toLin (root i) (coroot j))]
+    abel
+  suffices p.toLin.flip (coroot k) = p.toLin.flip (coroot l) from p.bijectiveRight.injective this
+  have _i : NoZeroSMulDivisors ℤ M := NoZeroSMulDivisors.int_of_charZero R M
+  have := injOn_dualMap_subtype_span_range_range (finite_range root)
+    (c := p.toLin.flip ∘ coroot) hp hr
+  apply this (mem_range_self k) (mem_range_self l)
+  refine Dual.eq_of_preReflection_mapsTo' hk₀ (finite_range root)
+    (Submodule.subset_span <| mem_range_self k) (hp k) (hr k) hkl ?_
+  rw [comp_apply, hl, hk, hij]
+  exact (hr i).comp <| (hr j).comp (hr i)
+
+/-- In characteristic zero if there is no torsion, to check that two finite familes of of roots and
+coroots form a root pairing, it is sufficient to check that they are stable under reflections. -/
+def mk' [Finite ι] [CharZero R] [NoZeroSMulDivisors R M]
+    (p : PerfectPairing R M N)
+    (root : ι ↪ M)
+    (coroot : ι ↪ N)
+    (hp : ∀ i, p.toLin (root i) (coroot i) = 2)
+    (hr : ∀ i, MapsTo (preReflection (root i) (p.toLin.flip (coroot i))) (range root) (range root))
+    (hc : ∀ i, MapsTo (preReflection (coroot i) (p.toLin (root i))) (range coroot) (range coroot)) :
+    RootPairing ι R M N where
+  toPerfectPairing := p
+  root := root
+  coroot := coroot
+  root_coroot_two := hp
+  reflection_perm i := equiv_of_mapsTo p root coroot i hr hp
+  reflection_perm_root i j := by
+    rw [equiv_of_mapsTo_apply, (exist_eq_reflection_of_mapsTo  p root coroot i j hr).choose_spec,
+      preReflection_apply, LinearMap.flip_apply]
+  reflection_perm_coroot i j := by
+    refine (coroot_eq_coreflection_of_root_eq' p root coroot hp hr hc ?_).symm
+    rw [equiv_of_mapsTo_apply, (exist_eq_reflection_of_mapsTo  p root coroot i j hr).choose_spec]
```

Comment:
Adapting to the `protected` above. And fixing some double spaces.
```suggestion
  reflection_perm i := RootPairing.equiv_of_mapsTo p root coroot i hr hp
  reflection_perm_root i j := by
    rw [RootPairing.equiv_of_mapsTo_apply,
      (exist_eq_reflection_of_mapsTo p root coroot i j hr).choose_spec,
      preReflection_apply, LinearMap.flip_apply]
  reflection_perm_coroot i j := by
    refine (coroot_eq_coreflection_of_root_eq' p root coroot hp hr hc ?_).symm
    rw [RootPairing.equiv_of_mapsTo_apply,
      (exist_eq_reflection_of_mapsTo p root coroot i j hr).choose_spec]
```

Reply:
Thanks, done.